### PR TITLE
Support PCH & PTH files

### PR DESCRIPTION
**Pull request by matthewtff**
Use contents of PCH and PTH files to calculate hash of direct cache
entry.

### CODE REVIEW

File: src/client/clang_command.cc [hunk 1]
```diff
@@ -100,6 +100,11 @@ bool ClangCommand::FillFlags(base::proto::Flags* flags,
              arg->getOption().matches(OPT_D) ||
              arg->getOption().matches(OPT_I)) {
       arg->render(arg_list_, non_cached_list);
+    } else if (arg->getOption().matches(OPT_include_pch) ||
+               arg->getOption().matches(OPT_include_pth)) {
+      auto* included_file = flags->add_included_files();
```
Comment (abyss7) [inline src/client/clang_command.cc]: Maybe just `flags->add_included_files(arg->getValue())` ?

File: src/client/clang_command.cc [hunk 2]
```diff
@@ -100,6 +100,11 @@ bool ClangCommand::FillFlags(base::proto::Flags* flags,
...
+      *included_file = arg->getValue();
+      arg->render(arg_list_, non_cached_list);
```
Comment (abyss7) [inline src/client/clang_command.cc]: Why to add to direct cache an explicit argument? - It's already taken into account in flags as `included_files`. You should render them later, like `sanitize_blacklist` or `deps_file`.

File: src/client/command.cc [hunk 3]
```diff
@@ -97,9 +97,8 @@ bool Command::GenerateFromArgs(int argc, const char* const raw_argv[],
     // TODO: move this code inside |DriverCommand| constructor.
     if ((command.getSource().getKind() != Action::AssembleJobClass &&
          command.getSource().getKind() != Action::BackendJobClass &&
-         command.getSource().getKind() != Action::CompileJobClass &&
-         command.getSource().getKind() != Action::PrecompileJobClass &&
```
Comment (abyss7) [inline src/client/command.cc]: Why do we skip the precompile jobs?

File: src/client/clang_command.cc [hunk 4]
```diff
@@ -100,6 +100,11 @@ bool ClangCommand::FillFlags(base::proto::Flags* flags,
              arg->getOption().matches(OPT_D) ||
              arg->getOption().matches(OPT_I)) {
       arg->render(arg_list_, non_cached_list);
+    } else if (arg->getOption().matches(OPT_include_pch) ||
+               arg->getOption().matches(OPT_include_pth)) {
+      auto* included_file = flags->add_included_files();
+      *included_file = arg->getValue();
```
Comment (abyss7) [inline src/client/clang_command.cc]: What are the rules for path calculation for a value of this option? Is it relative to current folder or input file? Or does it use include paths?

Comment (matthewtff) [inline src/client/clang_command.cc]: It is calculated relative to current work directory.

File: src/client/clang_command.cc (hunk 2 above)
Comment (matthewtff) [inline src/client/clang_command.cc]: This flag could be useful for preprocessing the source code, for example.
Why bother with manual rendering?

File: src/client/command.cc (hunk 3 above)
Comment (matthewtff) [inline src/client/command.cc]: As those work wrong on remotes. It looks like a PCH file generated on remote host doesn't work locally. There're may be several different reasons for this:
1. OS/Arch differencies(by order?) as pch/pth files are somethig like a memory dump.
2. These files contain absolute paths to source header.

File: src/client/clang_command.cc (hunk 2 above)
Comment (abyss7) [inline src/client/clang_command.cc]: Even without manual rendering still I don't see the point to render it inside `non_cached_list` - it's already counted inside `manifest.headers` for direct cache, and `non_cached_list` is not applied for simple cache.

File: src/cache/file_cache_test.cc [hunk 5]
```diff
@@ -496,6 +496,53 @@ TEST(FileCacheTest, DirectEntry_ChangedHeaderContents) {
...
 }
 
+TEST(FileCacheTest, DirectEntry_ChangedPreprocessedHeaderContents) {
+  const base::TemporaryDir tmp_dir;
+  const String path = tmp_dir;
+  const String object_path = path + "/test.o";
+  const String deps_path = path + "/test.d";
+  const String header1_path = path + "/test1.h";
+  const String header2_path = path + "/test2.h";
+  const String preprocessed_header_path = path + "/precompile.h";
```
Comment (abyss7) [inline src/cache/file_cache_test.cc]: Should use real-life extension, like .pch or .pth

**Comment by matthewtff** [comment]
Updated EmitterTest and used real life extensions.
https://github.com/abyss7/dist-clang/blob/master/src/cache/file_cache_test.cc#L31 makes no sense for preprocessed headers as they don't play in hash computation.

**Pull request merged**